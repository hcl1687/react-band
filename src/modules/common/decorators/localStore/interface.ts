import { IRBContext } from '~/interface'

export interface ILocalState {
  [propName: string]: any
}

export interface ILocalActionFactory {
  (...args: Array<any>): any
}

export interface ILocalActionFactories {
  [propName: string]: ILocalActionFactory
}

export interface ILocalAction {
  type: string
  payload?: any
}

export interface ILocalReducer {
  next: (state: ILocalState, action: ILocalAction) => ILocalState
  throw: (state: ILocalState, action: ILocalAction) => ILocalState
}

export interface ILocalReducers {
  [propName: string]: ILocalReducer
}

export interface ILocalStore {
  state: ILocalState
  actions: ILocalActionFactories
  reducers: ILocalReducers
}

export interface ILocalStoreFactory {
  (RB_CONTEXT: IRBContext): ILocalStore
}
