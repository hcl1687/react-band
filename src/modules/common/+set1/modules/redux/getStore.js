import { applyMiddleware, combineReducers, createStore } from 'redux'
import middlewares from './middleware'

const staticReducers = {
  // fix error:
  // Store does not have a valid reducer. Make sure the argument passed to combineReducers is an object whose values are reducers.
  '@ini': (state = {}) => (state)
}
let store

function createReducer (asyncReducers) {
  return combineReducers({
    ...staticReducers,
    ...asyncReducers
  })
}

export default function getStore (initialState = {}) {
  if (store) {
    return store
  }

  store = createStore(createReducer(), initialState, applyMiddleware(
    ...middlewares
  ))
  store.asyncReducers = {}

  store.injectReducer = (key, asyncReducer) => {
    if (!key || store.asyncReducers[key]) {
      return
    }

    store.asyncReducers[key] = asyncReducer
    store.replaceReducer(createReducer(store.asyncReducers))
  }

  return store
}
