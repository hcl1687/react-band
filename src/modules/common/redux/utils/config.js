export default () => {
  return {
    name: 'reduxUtils'
  }
}
