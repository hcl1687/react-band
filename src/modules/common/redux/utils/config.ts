import { IRBLeafConfig } from '~/interface'

export default (): IRBLeafConfig => {
  return {
    name: 'reduxUtils'
  }
}
