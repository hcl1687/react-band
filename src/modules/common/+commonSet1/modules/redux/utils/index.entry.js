import getStore from '../getStore'

const entry = () => {
  return {
    getStore
  }
}

export default {
  entry
}
