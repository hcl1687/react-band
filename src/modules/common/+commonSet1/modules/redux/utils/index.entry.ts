import getStore from '../getStore'

export default (): RB.IRBModule => {
  return {
    getStore
  }
}
