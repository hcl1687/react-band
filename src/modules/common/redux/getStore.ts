import { applyMiddleware, combineReducers, createStore } from 'redux'
import middlewares from './middleware'

export interface IState {
  [propName: string]: any
}

export interface IReducer {
  (state: IState, payload?: any): IState
}

const staticReducers = {
  // fix error:
  // Store does not have a valid reducer. Make sure the argument passed to combineReducers is an object whose values are reducers.
  '@ini': (state = {}) => (state)
}
let store: IState

function createReducer (asyncReducers: { [propName: string]: IReducer } = {}) {
  return combineReducers({
    ...staticReducers,
    ...asyncReducers
  })
}

export default function getStore (initialState = {}): IState {
  if (store) {
    return store
  }

  store = createStore(createReducer(), initialState, applyMiddleware(
    ...middlewares
  ))
  store.asyncReducers = {}

  store.injectReducer = (key: string, asyncReducer: IReducer) => {
    if (!key || store.asyncReducers[key]) {
      return
    }

    store.asyncReducers[key] = asyncReducer
    store.replaceReducer(createReducer(store.asyncReducers))
  }

  return store
}
