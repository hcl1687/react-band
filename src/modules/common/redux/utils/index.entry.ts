import { IRBModule } from '~/interface'
import getStore from '../getStore'

export default (): IRBModule => {
  return {
    getStore
  }
}
