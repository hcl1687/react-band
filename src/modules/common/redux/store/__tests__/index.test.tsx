import PropTypes, { InferProps } from 'prop-types'
import React, { Component } from 'react'
import { mount, render } from 'enzyme'
import { IRBContext } from '~/interface'
import { Provider } from 'react-redux'
import getStore from '../../getStore'
import storeDecoFactory from '../index.entry'
import tools from '~/../tests/utils/index'
import utils from '~/../tests/utils/mockUtils'

const store = getStore()

class Test extends Component<InferProps<typeof Test.propTypes>> {
  static propTypes = {
    increase: PropTypes.func,
    decrease: PropTypes.func,
    count: PropTypes.number
  }

  componentDidMount () {
    const { increase } = this.props
    increase && increase()
  }

  handleAdd = () => {
    const { increase } = this.props
    increase()
  }

  handleSubstract = () => {
    const { decrease } = this.props
    decrease()
  }

  render () {
    return <div className='test'>{this.props.count}</div>
  }
}

interface IState {
  [propName: string]: any
}

const state = {
  count: 0
}
const actions = {
  increase: (data: number) => data,
  decrease: (data: number) => data
}
const reducers = {
  increase: {
    next (state: IState) {
      const { count } = state
      return {
        ...state,
        count: count + 1
      }
    },
    throw (state: IState) {
      return state
    }
  },
  decrease: {
    next (state: IState) {
      const { count } = state
      return {
        ...state,
        count: count - 1
      }
    },
    throw (state: IState) {
      return state
    }
  }
}

function mathStoreFactory () {
  return {
    state,
    actions,
    reducers
  }
}

const context: IRBContext = {
  options: {},
  modules: {},
  i18ns: {},
  themes: {},
  packedModules: {},
  modulesConfig: {},
  routes: [],
  getModule: async (key: string) => {
    if (key === 'utils') {
      return utils
    } else if (key === 'mathStore') {
      return mathStoreFactory()
    }
  }
}

describe('common/redux/store', () => {
  it('should render correctly', async () => {
    const StoreDeco = await storeDecoFactory(context)
    const targetConfig = {
      name: 'test',
      decoratorsConfig: {
        '@reduxStore': {
          mathStore: {
            actions: ['increase', 'decrease'],
            state: ['count']
          }
        }
      }
    }

    const TestWithStoreDeco = await StoreDeco(targetConfig, { name: '@reduxStore' }, context)(Test)
    const wrapper = render(
      <Provider store={store}>
        <TestWithStoreDeco />
      </Provider>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should mount correctly', async () => {
    const storeDeco = await storeDecoFactory(context)
    const targetConfig = {
      name: 'test',
      decoratorsConfig: {
        '@reduxStore': {
          mathStore: {
            actions: ['increase', 'decrease'],
            state: ['count']
          }
        }
      }
    }

    const TestWithStoreDeco = await storeDeco(targetConfig, { name: '@reduxStore' }, context)(Test)
    const wrapper = mount(
      <Provider store={store}>
        <TestWithStoreDeco />
      </Provider>
    )

    await tools.delay(() => {
      expect(wrapper.find('.test').text()).toEqual('1')
    })
  })
})

describe('common/redux/store/production', () => {
  const NODE_ENV = process.env.NODE_ENV
  beforeAll(() => {
    process.env.NODE_ENV = 'production'
  })
  afterAll(() => {
    process.env.NODE_ENV = NODE_ENV
  })
  it('should render correctly', async () => {
    const StoreDeco = await storeDecoFactory(context)
    const targetConfig = {
      name: 'test',
      decoratorsConfig: {
        '@reduxStore': {
          mathStore: {
            actions: ['increase', 'decrease'],
            state: ['count']
          }
        }
      }
    }

    const TestWithStoreDeco = await StoreDeco(targetConfig, { name: '@reduxStore' }, context)(Test)
    const wrapper = render(
      <Provider store={store}>
        <TestWithStoreDeco />
      </Provider>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
