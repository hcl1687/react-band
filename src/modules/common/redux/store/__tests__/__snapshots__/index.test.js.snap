// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/redux/store should render correctly 1`] = `
<div
  class="test"
>
  0
</div>
`;

exports[`common/redux/store/production should render correctly 1`] = `
<div
  class="test"
>
  1
</div>
`;
